### PR TITLE
AbsPkt improvements

### PR DESCRIPTION
**Pull request by mlimbeck**
- moved `absInfofield` functions from `hopfield_spec.gobra`to` infofield_spec.gobra`
- added opaque to `absPkt` and subsequent functions

### CODE REVIEW

File: pkg/slayers/path/hopfield_spec.gobra [hunk 1]
```diff
@@ -16,32 +16,6 @@
...
-decreases
-pure func ConsDir(raw []byte, currINF int) bool {
-	return raw[InfoFieldOffset(currINF)] & 0x1 == 0x1
-}
-
-ghost 
-requires 0 <= currINF
-requires InfoFieldOffset(currINF) < len(raw)
-requires acc(&raw[InfoFieldOffset(currINF)], _)
-decreases
```
Comment (jcp19) [inline pkg/slayers/path/hopfield_spec.gobra:44]: Nice catch! Thanks for moving this!

File: router/io-spec.gobra [hunk 2]
```diff
@@ -469,25 +480,15 @@ pure func midSeg(
...
-	let base := scion.Base{metaHdr,
-		numInfoFields(seg1, seg2, seg3),
-		seg1 + seg2 + seg3} in
-	metaHdr.InBounds() &&
-	0 < metaHdr.SegLen[0] &&
-	base.ValidCurrInfSpec() &&
-	base.ValidCurrHfSpec() &&
-	len(raw) >= pktLen(seg1, seg2, seg3)
+requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R55)
+requires validPktMetaHdr(raw)
```
Comment (jcp19) [inline router/io-spec.gobra:484]: This is a good idea

File: router/io-spec.gobra [hunk 3]
```diff
@@ -469,25 +480,15 @@ pure func midSeg(
...
+//ensures res != none[io.IO_pkt2] ==> len(get(res).CurrSeg.Future) > 0
+//ensures res != none[io.IO_pkt2] ==> get(res).CurrSeg.Future[0].extr_asid() == dp.Asid()
```
Comment (jcp19) [inline router/io-spec.gobra:486]: Why are these posts here? Is the goal to prove them before we merge? Maybe sth that might make sense here is to prove these conditions extrinsically (i.e., on a separate lemma) and call the lemma when we need them.

**Review by jcp19 — APPROVED** [review]
This looks all fine to me. But didn't we talk about merging this in master instead?

File: router/io-spec.gobra (hunk 3 above)
Comment (mlimbeck) [reply router/io-spec.gobra]: Yes if we want to keep everything opaque, we have to prove them separately. I removed them for now.

**Review by Dspil — APPROVED** [review]

**Comment by jcp19** [comment]
it seems that this PR introduced a new verification error. @mlimbeck can you fix it before we merge?

**Comment by jcp19** [comment]
Interestingly, the last commit performs changes in the same lines as PR #276, which @Dspil introduced so that he could make progress with the unsupported paths. I guess both postconditions make sense, but we should discuss briefly over a meeting which ones are the most useful. We can merge this PR after we clarify that, hopefully tomorrow.

File: router/io-spec.gobra [hunk 4]
```diff
@@ -190,10 +191,13 @@ requires 0 <= seg3Len
 requires hopFieldOffset(numINF, seg1Len + seg2Len + seg3Len) <= len(raw)
 requires currINFIdx <= numINF + 1
 requires 1 <= currINFIdx && currINFIdx < 4
-requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _)
+requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R56)
+ensures (currINFIdx == 1 && seg2Len > 0 && res != none[seq[io.IO_as]]) ==> len(get(res)) == seg2Len
+ensures (currINFIdx == 2 && seg2Len > 0 && seg3Len > 0 && res != none[seq[io.IO_as]]) ==> len(get(res)) == seg3Len
```
Comment (jcp19) [inline router/io-spec.gobra:196]: could we have the postconditions proven in a separate lemma?

File: router/io-spec.gobra [hunk 5]
```diff
@@ -235,31 +240,34 @@ requires 0 <= seg3Len
 requires hopFieldOffset(numINF, seg1Len + seg2Len + seg3Len) <= len(raw)
 requires currINFIdx <= numINF + 1
 requires 2 <= currINFIdx && currINFIdx < 5
-requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _)
+requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R56)
 requires (currINFIdx == 4 && seg2Len > 0) ==> asid != none[io.IO_as]
 requires (currINFIdx == 2 && seg2Len > 0 && seg3Len > 0) ==> asid != none[io.IO_as]
+ensures	 (currINFIdx == 4 && seg2Len > 0 && res != none[seq[io.IO_as]]) ==> len(get(res)) == seg1Len
+ensures	 (currINFIdx == 2 && seg2Len > 0 && seg3Len > 0 && res != none[seq[io.IO_as]]) ==> len(get(res)) == seg3Len
```
Comment (jcp19) [inline router/io-spec.gobra:247]: same here

File: router/io-spec.gobra [hunk 6]
```diff
@@ -469,25 +480,13 @@ pure func midSeg(
...
-		numInfoFields(seg1, seg2, seg3),
-		seg1 + seg2 + seg3} in
-	metaHdr.InBounds() &&
-	0 < metaHdr.SegLen[0] &&
-	base.ValidCurrInfSpec() &&
-	base.ValidCurrHfSpec() &&
-	len(raw) >= pktLen(seg1, seg2, seg3)
+requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R56)
+requires validPktMetaHdr(raw)
+opaque
```
Comment (jcp19) [inline router/io-spec.gobra:485]: as a general guideline, we should put the `opaque` modifier in the line right after `ghost` to make it easier to read.

**Review by jcp19 — APPROVED** [review]

File: router/io-spec.gobra (hunk 6 above)
Comment (mlimbeck) [reply router/io-spec.gobra]: Okay, I changed it.

**Pull request merged**